File: apps/desktop/src/lib/ipc/library.ts
// Mirror of `crates/oxyn-desktop/src/ipc/library.rs`, as executable schemas:
// both sides still change in the same commit, but a field renamed on one side
// only is now named at the boundary instead of read as `undefined` (ADR-0031).

import { z } from "zod"

import { call, Nothing } from "./client"
import { ResultColumn } from "./types"

export const DocumentChange = z.object({
  document: z.string(),
  /** Monotonic per document; the backend refuses one that does not advance. */
  revision: z.number().int().nonnegative(),
  title: z.string(),
  text: z.string(),
  connection: z.string().nullable(),
  named: z.boolean(),
})
export type DocumentChange = z.infer<typeof DocumentChange>

export const DocumentWrite = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("saved"),
    revision: z.number().int().nonnegative(),
    savedRevision: z.number().int().nonnegative(),
    isSaved: z.boolean(),
  }),
  z.object({ type: z.literal("closed") }),
  /** The stored document moved: keep the local text, offer a new copy. */
  z.object({ type: z.literal("conflict"), message: z.string() }),
  /** A newer draft replaced this one before it was written. */
  z.object({ type: z.literal("superseded") }),
])
export type DocumentWrite = z.infer<typeof DocumentWrite>

export const DocumentView = z.object({
  id: z.string(),
  title: z.string(),
  text: z.string(),
  savedTitle: z.string().nullable(),
  savedText: z.string().nullable(),
  revision: z.number().int().nonnegative(),
  savedRevision: z.number().int().nonnegative(),
  isSaved: z.boolean(),
  isOpen: z.boolean(),
  connection: z.string().nullable(),
  fromAgent: z.boolean(),
})
export type DocumentView = z.infer<typeof DocumentView>

export const DocumentQuery = z.object({
  savedOnly: z.boolean(),
  openOnly: z.boolean(),
  search: z.string(),
  before: z.string().nullable(),
  limit: z.number().int().nonnegative().nullable(),
})
export type DocumentQuery = z.infer<typeof DocumentQuery>

export const DocumentEntry = z.object({
  id: z.string(),
  title: z.string(),
  connection: z.string().nullable(),
  connectionName: z.string().nullable(),
  /** RFC 3339, UTC: the backend formats it, the front never reparses it. */
  updatedAt: z.string(),
  isSaved: z.boolean(),
  isOpen: z.boolean(),
  hasChanges: z.boolean(),
  fromAgent: z.boolean(),
})
export type DocumentEntry = z.infer<typeof DocumentEntry>

export const DocumentList = z.object({
  entries: z.array(DocumentEntry),
  next: z.string().nullable(),
})
export type DocumentList = z.infer<typeof DocumentList>

export const HistoryStatusChoice = z.enum([
  "all",
  "running",
  "succeeded",
  "failed",
  "cancelled",
  "denied",
  "ambiguous",
])
export type HistoryStatusChoice = z.infer<typeof HistoryStatusChoice>

export const HistoryQuery = z.object({
  connection: z.string().nullable(),
  search: z.string(),
  /** `null` means all dates. */
  days: z.number().int().nonnegative().nullable(),
  status: HistoryStatusChoice,
  before: z.number().int().nullable(),
  limit: z.number().int().nonnegative().nullable(),
})
export type HistoryQuery = z.infer<typeof HistoryQuery>

export const HistoryRow = z.object({
  id: z.number().int(),
  /** RFC 3339, UTC: the backend formats it, the front never reparses it. */
  at: z.string(),
  connectionName: z.string().nullable(),
  /** At most 256 characters; never executed. */
  preview: z.string(),
  status: z.string(),
  durationMs: z.number().nonnegative().nullable(),
  rows: z.number().int().nonnegative().nullable(),
  needsInspection: z.boolean(),
  /** The user declared this write inspected on the server; it no longer warns. */
  reconciled: z.boolean(),
  /** The connection it ran on, to address a retained result. Never rendered. */
  connection: z.string().nullable(),
  /** A result that may still be retained; its buffer can have expired. */
  result: z.string().nullable(),
  /** Submitted by an agent: a copy keeps saying so (ADR-0023). */
  fromAgent: z.boolean(),
})
export type HistoryRow = z.infer<typeof HistoryRow>

export const HistoryList = z.object({
  entries: z.array(HistoryRow),
  next: z.number().int().nullable(),
})
export type HistoryList = z.infer<typeof HistoryList>

export const HistoryDetail = z.object({
  id: z.number().int(),
  statement: z.string(),
  connectionName: z.string().nullable(),
  status: z.string(),
  error: z.string().nullable(),
  needsInspection: z.boolean(),
  /** Submitted by an agent: the console it is copied into says so. */
  fromAgent: z.boolean(),
})
export type HistoryDetail = z.infer<typeof HistoryDetail>

export const HistoryConnection = z.object({
  connection: z.string(),
  name: z.string().nullable(),
  inWorkspace: z.boolean(),
})
export type HistoryConnection = z.infer<typeof HistoryConnection>

export const HistoryConnectionList = z.object({
  entries: z.array(HistoryConnection),
  next: z.number().int().nullable(),
})
export type HistoryConnectionList = z.infer<typeof HistoryConnectionList>

export const RetainedResult = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("open"),
    result: z.string(),
    columns: z.array(ResultColumn),
    rows: z.number().int().nonnegative(),
    complete: z.boolean(),
    /** The buffer stopped retaining rows at its bound. */
    truncated: z.boolean(),
  }),
  /** Released: nothing is rerun to bring it back. */
  z.object({ type: z.literal("expired") }),
])
export type RetainedResult = z.infer<typeof RetainedResult>

export const library = {
  /** Minted by the backend: library pages order documents by identifier age. */
  newDocument: () => call("new_query_document", z.string()),

  /** `commandId` lets `backend.cancel` stop the write before it commits. */
  saveDocument: (commandId: string, change: DocumentChange) =>
    call("save_query_document", DocumentWrite, { commandId, change }),

  /** `commandId` lets `backend.cancel` stop the close before it commits. */
  closeDocument: (
    commandId: string,
    document: string,
    revision: number,
    discard: boolean
  ) =>
    call("close_query_document", DocumentWrite, {
      commandId,
      document,
      revision,
      discard,
    }),

  releaseDocument: (document: string) =>
    call("release_query_document", Nothing, { document }),

  deleteDocument: (document: string, revision: number) =>
    call("delete_query_document", Nothing, { document, revision }),

  openDocument: (document: string) =>
    call("open_query_document", DocumentView, { document }),

  listDocuments: (commandId: string, query: DocumentQuery) =>
    call("list_query_documents", DocumentList, { commandId, query }),

  readHistory: (commandId: string, query: HistoryQuery) =>
    call("read_history", HistoryList, { commandId, query }),

  readHistoryEntry: (entry: number) =>
    call("read_history_entry", HistoryDetail, { entry }),

  /** Declares an unresolved write inspected on the server. Retries nothing. */
  reconcileHistoryEntry: (entry: number) =>
    call("reconcile_history_entry", Nothing, { entry }),

  /** Reopens a retained buffer: no query, no session. */
  openRetainedResult: (connection: string, result: string) =>
    call("open_retained_result", RetainedResult, { connection, result }),

  historyConnections: (before: number | null) =>
    call("list_history_connections", HistoryConnectionList, { before }),
}

File: crates/oxyn-desktop/src/ipc/library.rs
//! What crosses the boundary for query documents and local history.
//!
//! Lists carry no complete SQL bodies: a page is bounded, and a body is read
//! only when one entry is opened ([ADR-0014](../../../../docs/adr/0014-documents-et-historique.md)).
//! Bound values never appear here — they are never stored.

use oxyn_core::HistoryStatusFilter;
use oxyn_store::ActorKind;
use oxyn_store::documents::{DocumentPage, DocumentSummary};
use oxyn_store::history::{HistoryConnectionPage, HistoryPage, HistorySummary};
use serde::{Deserialize, Serialize};

/// A working or named copy change, as a console submits it.
#[derive(Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentChange {
    pub document: String,
    /// Monotonic per document, chosen by the console: the backend refuses a
    /// revision that does not advance.
    pub revision: u64,
    pub title: String,
    pub text: String,
    /// The connection the text was written for; it does not connect anything.
    pub connection: Option<String>,
    /// Also write the named copy the library lists.
    #[serde(default)]
    pub named: bool,
}

// Not derived: the text is the user's SQL, which a log line has no reason to
// carry whole.
impl std::fmt::Debug for DocumentChange {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("DocumentChange")
            .field("revision", &self.revision)
            .field("text_bytes", &self.text.len())
            .field("named", &self.named)
            .finish_non_exhaustive()
    }
}

/// The answer to a save or a close.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum DocumentWrite {
    /// The store acknowledged exactly what was sent.
    #[serde(rename_all = "camelCase")]
    Saved {
        revision: u64,
        saved_revision: u64,
        is_saved: bool,
    },
    Closed,
    /// The stored document moved under this console. Nothing local is lost:
    /// the console keeps its text and offers a new copy
    /// ([ADR-0016](../../../../docs/adr/0016-autosauvegarde-bornee.md)).
    Conflict {
        message: String,
    },
    /// A newer draft replaced this one before it was written, or the user
    /// cancelled: not a failure.
    Superseded,
}

/// A document opened in full.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentView {
    pub id: String,
    pub title: String,
    pub text: String,
    pub saved_title: Option<String>,
    pub saved_text: Option<String>,
    pub revision: u64,
    pub saved_revision: u64,
    pub is_saved: bool,
    pub is_open: bool,
    pub connection: Option<String>,
    pub from_agent: bool,
}

impl From<oxyn_store::Document> for DocumentView {
    fn from(document: oxyn_store::Document) -> Self {
        Self {
            id: document.id.to_string(),
            title: document.title,
            text: document.content,
            saved_title: document.saved_title,
            saved_text: document.saved_content,
            revision: document.revision,
            saved_revision: document.saved_revision,
            is_saved: document.is_saved,
            is_open: document.is_open,
            connection: document.connection.map(|id| id.to_string()),
            from_agent: document.provenance.is_some(),
        }
    }
}

/// Which documents a library page lists.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentQuery {
    #[serde(default)]
    pub saved_only: bool,
    #[serde(default)]
    pub open_only: bool,
    #[serde(default)]
    pub search: String,
    pub before: Option<String>,
    pub limit: Option<u16>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentEntry {
    pub id: String,
    pub title: String,
    pub connection: Option<String>,
    /// Absent when the connection was removed.
    pub connection_name: Option<String>,
    /// RFC 3339, UTC.
    pub updated_at: String,
    pub is_saved: bool,
    pub is_open: bool,
    pub has_changes: bool,
    pub from_agent: bool,
}

impl From<DocumentSummary> for DocumentEntry {
    fn from(summary: DocumentSummary) -> Self {
        Self {
            id: summary.id.to_string(),
            title: summary.title,
            connection: summary.connection.map(|id| id.to_string()),
            connection_name: summary.connection_name,
            updated_at: summary.updated_at.to_rfc3339(),
            is_saved: summary.is_saved,
            is_open: summary.is_open,
            has_changes: summary.has_changes,
            from_agent: summary.from_agent,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentList {
    pub entries: Vec<DocumentEntry>,
    pub next: Option<String>,
}

impl From<DocumentPage> for DocumentList {
    fn from(page: DocumentPage) -> Self {
        Self {
            entries: page.entries.into_iter().map(Into::into).collect(),
            next: page.next.map(|id| id.to_string()),
        }
    }
}

/// Which executions a history page lists.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryQuery {
    pub connection: Option<String>,
    #[serde(default)]
    pub search: String,
    /// `None` means all dates.
    pub days: Option<u16>,
    #[serde(default)]
    pub status: HistoryStatusChoice,
    pub before: Option<i64>,
    pub limit: Option<u16>,
}

#[derive(Debug, Clone, Copy, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum HistoryStatusChoice {
    #[default]
    All,
    Running,
    Succeeded,
    Failed,
    Cancelled,
    Denied,
    Ambiguous,
}

impl From<HistoryStatusChoice> for HistoryStatusFilter {
    fn from(choice: HistoryStatusChoice) -> Self {
        match choice {
            HistoryStatusChoice::All => Self::All,
            HistoryStatusChoice::Running => Self::Running,
            HistoryStatusChoice::Succeeded => Self::Succeeded,
            HistoryStatusChoice::Failed => Self::Failed,
            HistoryStatusChoice::Cancelled => Self::Cancelled,
            HistoryStatusChoice::Denied => Self::Denied,
            HistoryStatusChoice::Ambiguous => Self::Ambiguous,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryRow {
    pub id: i64,
    /// RFC 3339, UTC.
    pub at: String,
    pub connection_name: Option<String>,
    /// At most 256 characters; never executed.
    pub preview: String,
    pub status: String,
    pub duration_ms: Option<u64>,
    pub rows: Option<u64>,
    /// An ambiguous or unresolved write: inspect the server, never replay
    /// ([I-13](../../../../CLAUDE.md#i-13)).
    pub needs_inspection: bool,
    /// The user declared this write inspected on the server; it no longer warns.
    pub reconciled: bool,
    /// The connection it ran on, to address a retained result. Never rendered.
    pub connection: Option<String>,
    /// A result that may still be retained; its buffer can have expired.
    pub result: Option<String>,
    /// Submitted by an agent: a copy keeps saying so
    /// ([ADR-0023](../../../../docs/adr/0023-fournisseurs-declares-et-provenance.md)).
    pub from_agent: bool,
}

impl From<HistorySummary> for HistoryRow {
    fn from(summary: HistorySummary) -> Self {
        Self {
            id: summary.id,
            at: summary.ts.to_rfc3339(),
            connection_name: summary.connection_name,
            preview: summary.statement_preview,
            status: summary.status.as_str().to_owned(),
            duration_ms: summary.duration.map(crate::ipc::millis),
            rows: summary.rows,
            needs_inspection: summary.requires_reconciliation,
            reconciled: summary.reconciled_at.is_some(),
            connection: summary.connection.map(|id| id.to_string()),
            result: summary.result.map(|id| id.to_string()),
            from_agent: summary.actor_kind == ActorKind::Agent,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryList {
    pub entries: Vec<HistoryRow>,
    pub next: Option<i64>,
}

impl From<HistoryPage> for HistoryList {
    fn from(page: HistoryPage) -> Self {
        Self {
            entries: page.entries.into_iter().map(Into::into).collect(),
            next: page.next,
        }
    }
}

/// One execution, read in full to be copied into a console — never replayed.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryDetail {
    pub id: i64,
    pub statement: String,
    pub connection_name: Option<String>,
    pub status: String,
    pub error: Option<String>,
    pub needs_inspection: bool,
    /// Submitted by an agent: the console it is copied into says so.
    pub from_agent: bool,
}

impl From<oxyn_store::HistoryEntry> for HistoryDetail {
    fn from(entry: oxyn_store::HistoryEntry) -> Self {
        let needs_inspection = entry.record.requires_reconciliation();
        Self {
            id: entry.id,
            statement: entry.record.statement,
            connection_name: entry.record.connection_name,
            status: entry.record.status.as_str().to_owned(),
            error: entry.record.error,
            needs_inspection,
            from_agent: entry.record.actor_kind == ActorKind::Agent,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryConnection {
    pub connection: String,
    pub name: Option<String>,
    /// False once removed: its entries stay listed and searchable.
    pub in_workspace: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HistoryConnectionList {
    pub entries: Vec<HistoryConnection>,
    pub next: Option<i64>,
}

impl From<HistoryConnectionPage> for HistoryConnectionList {
    fn from(page: HistoryConnectionPage) -> Self {
        Self {
            entries: page
                .entries
                .into_iter()
                .map(|entry| HistoryConnection {
                    connection: entry.connection.to_string(),
                    name: entry.name,
                    in_workspace: entry.in_workspace,
                })
                .collect(),
            next: page.next,
        }
    }
}

/// A retained result reopened without running anything.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum RetainedResult {
    #[serde(rename_all = "camelCase")]
    Open {
        result: String,
        columns: Vec<crate::ipc::ResultColumn>,
        rows: usize,
        /// Only an exhausted stream describes a whole result.
        complete: bool,
        /// The buffer stopped retaining rows at its bound.
        truncated: bool,
    },
    /// The buffer was released: nothing is rerun to bring it back.
    Expired,
}
